File: ios/Plugin/CapacitorPluginOcr.swift
import Capacitor
import Foundation
import ImageIO
import Photos
import UIKit
import Vision

@objc(CapacitorPluginOcr)
public class CapacitorPluginOcr: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "CapacitorPluginOcr"
    public let jsName = "CapacitorPluginOcr"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(#selector(checkPermissions(_:)), returnType: .promise),
        CAPPluginMethod(#selector(requestPermissions(_:)), returnType: .promise),
        CAPPluginMethod(#selector(recognizeEnglishText(_:)), returnType: .promise),
        CAPPluginMethod(#selector(cropImage(_:)), returnType: .promise),
        CAPPluginMethod(#selector(startCropUI(_:)), returnType: .promise)
    ]

    @objc public override func checkPermissions(_ call: CAPPluginCall) {
        call.resolve(["granted": true])
    }

    @objc public override func requestPermissions(_ call: CAPPluginCall) {
        call.resolve(["granted": true])
    }

    @objc public func recognizeEnglishText(_ call: CAPPluginCall) {
        guard let imagePath = call.getString("imagePath"), !imagePath.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
            call.reject("Image path is required")
            return
        }

        let cropArea = call.getObject("cropArea")
        loadImage(from: imagePath) { [weak self] image, errorMessage in
            guard let self = self else { return }

            if let errorMessage = errorMessage {
                call.reject(errorMessage)
                return
            }

            guard var workingImage = image else {
                call.reject("Could not load image")
                return
            }

            if let cropArea = cropArea {
                guard let croppedImage = self.crop(image: workingImage, cropArea: cropArea) else {
                    call.reject("Could not crop image")
                    return
                }
                workingImage = croppedImage
            }

            self.performOCR(on: workingImage, call: call)
        }
    }

    @objc public func cropImage(_ call: CAPPluginCall) {
        guard let imagePath = call.getString("imagePath"), !imagePath.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
            call.reject("Image path is required")
            return
        }

        guard let cropArea = call.getObject("cropArea") else {
            call.reject("Crop area is required")
            return
        }

        loadImage(from: imagePath) { [weak self] image, errorMessage in
            guard let self = self else { return }

            if let errorMessage = errorMessage {
                call.reject(errorMessage)
                return
            }

            guard let image = image, let croppedImage = self.crop(image: image, cropArea: cropArea) else {
                call.reject("Could not crop image")
                return
            }

            do {
                let filePath = try self.writeTemporaryJpeg(croppedImage)
                call.resolve(["croppedImagePath": filePath])
            } catch {
                call.reject("Could not save cropped image: \(error.localizedDescription)")
            }
        }
    }

    @objc public func startCropUI(_ call: CAPPluginCall) {
        call.reject("Interactive crop UI is not implemented. Use cropImage with a cropArea instead.")
    }

    private func loadImage(from imagePath: String, completion: @escaping (UIImage?, String?) -> Void) {
        let trimmedPath = imagePath.trimmingCharacters(in: .whitespacesAndNewlines)

        if trimmedPath.hasPrefix("ph://") {
            loadPhotoAsset(identifier: String(trimmedPath.dropFirst(5)), completion: completion)
            return
        }

        if trimmedPath.hasPrefix("file://"), let url = URL(string: trimmedPath) {
            completion(UIImage(contentsOfFile: url.path), nil)
            return
        }

        if let url = URL(string: trimmedPath), url.scheme != nil, url.scheme != "file" {
            DispatchQueue.global(qos: .userInitiated).async {
                do {
                    let data = try Data(contentsOf: url)
                    let image = UIImage(data: data)
                    DispatchQueue.main.async {
                        completion(image, image == nil ? "Could not decode image data" : nil)
                    }
                } catch {
                    DispatchQueue.main.async {
                        completion(nil, "Could not load image: \(error.localizedDescription)")
                    }
                }
            }
            return
        }

        completion(UIImage(contentsOfFile: trimmedPath), nil)
    }

    private func loadPhotoAsset(identifier: String, completion: @escaping (UIImage?, String?) -> Void) {
        let fetchResult = PHAsset.fetchAssets(withLocalIdentifiers: [identifier], options: nil)
        guard let asset = fetchResult.firstObject else {
            completion(nil, "Could not find photo asset")
            return
        }

        let options = PHImageRequestOptions()
        options.deliveryMode = .highQualityFormat
        options.isNetworkAccessAllowed = true

        PHImageManager.default().requestImage(
            for: asset,
            targetSize: PHImageManagerMaximumSize,
            contentMode: .default,
            options: options
        ) { image, _ in
            DispatchQueue.main.async {
                completion(image, image == nil ? "Could not load photo asset" : nil)
            }
        }
    }

    private func performOCR(on image: UIImage, call: CAPPluginCall) {
        guard let cgImage = image.cgImage else {
            call.reject("Could not get CGImage from UIImage")
            return
        }

        let request = VNRecognizeTextRequest { [weak self] request, error in
            if let error = error {
                DispatchQueue.main.async {
                    call.reject("OCR failed: \(error.localizedDescription)")
                }
                return
            }

            let observations = request.results as? [VNRecognizedTextObservation] ?? []
            let result = self?.processOCRResults(observations) ?? ["words": [], "rawWords": []]
            DispatchQueue.main.async {
                call.resolve(result)
            }
        }

        request.recognitionLanguages = ["en-US"]
        request.recognitionLevel = .accurate
        request.usesLanguageCorrection = true

        let handler = VNImageRequestHandler(
            cgImage: cgImage,
            orientation: CGImagePropertyOrientation(image.imageOrientation),
            options: [:]
        )

        DispatchQueue.global(qos: .userInitiated).async {
            do {
                try handler.perform([request])
            } catch {
                DispatchQueue.main.async {
                    call.reject("Failed to perform OCR: \(error.localizedDescription)")
                }
            }
        }
    }

    private func processOCRResults(_ observations: [VNRecognizedTextObservation]) -> [String: Any] {
        let regex = try? NSRegularExpression(pattern: "[a-zA-Z]+", options: [])
        var uniqueWords = Set<String>()
        var rawWords: [String] = []
        var resultWords: [[String: Any]] = []

        for observation in observations {
            guard let candidate = observation.topCandidates(1).first else { continue }
            let text = candidate.string
            let range = NSRange(text.startIndex..<text.endIndex, in: text)

            regex?.matches(in: text, options: [], range: range).forEach { match in
                guard let wordRange = Range(match.range, in: text) else { return }

                let word = String(text[wordRange]).lowercased()
                if word.count <= 1 {
                    return
                }

                rawWords.append(word)
                if uniqueWords.insert(word).inserted {
                    let lemma = lemmatize(word)
                    var wordResult: [String: Any] = [
                        "word": word,
                        "confidence": Double(candidate.confidence)
                    ]

                    if lemma != word {
                        wordResult["lemma"] = lemma
                    }

                    resultWords.append(wordResult)
                }
            }
        }

        return [
            "words": resultWords,
            "rawWords": rawWords
        ]
    }

    private func crop(image: UIImage, cropArea: JSObject) -> UIImage? {
        guard let cgImage = image.cgImage else {
            return nil
        }

        let imageWidth = CGFloat(cgImage.width)
        let imageHeight = CGFloat(cgImage.height)
        let x = CGFloat(clamp(doubleValue(cropArea["x"], defaultValue: 0), min: 0, max: 1))
        let y = CGFloat(clamp(doubleValue(cropArea["y"], defaultValue: 0), min: 0, max: 1))
        let width = CGFloat(clamp(doubleValue(cropArea["width"], defaultValue: 1), min: 0, max: 1))
        let height = CGFloat(clamp(doubleValue(cropArea["height"], defaultValue: 1), min: 0, max: 1))

        let cropX = min(x * imageWidth, imageWidth - 1)
        let cropY = min(y * imageHeight, imageHeight - 1)
        let cropWidth = max(1, min(width * imageWidth, imageWidth - cropX))
        let cropHeight = max(1, min(height * imageHeight, imageHeight - cropY))
        let cropRect = CGRect(x: cropX, y: cropY, width: cropWidth, height: cropHeight)

        guard let croppedCGImage = cgImage.cropping(to: cropRect) else {
            return nil
        }

        return UIImage(cgImage: croppedCGImage, scale: image.scale, orientation: image.imageOrientation)
    }

    private func writeTemporaryJpeg(_ image: UIImage) throws -> String {
        guard let data = image.jpegData(compressionQuality: 0.92) else {
            throw OcrPluginError.imageEncodingFailed
        }

        let fileURL = URL(fileURLWithPath: NSTemporaryDirectory())
            .appendingPathComponent("ocr_crop_\(Int(Date().timeIntervalSince1970 * 1000)).jpg")
        try data.write(to: fileURL, options: .atomic)
        return fileURL.path
    }

    private func doubleValue(_ value: Any?, defaultValue: Double) -> Double {
        if let double = value as? Double {
            return double
        }

        if let number = value as? NSNumber {
            return number.doubleValue
        }

        return defaultValue
    }

    private func clamp(_ value: Double, min: Double, max: Double) -> Double {
        Swift.max(min, Swift.min(max, value))
    }

    private func lemmatize(_ word: String) -> String {
        let lowerWord = word.lowercased()
        let irregularVerbs: [String: String] = [
            "was": "be", "were": "be", "been": "be", "being": "be",
            "had": "have", "has": "have",
            "did": "do", "done": "do",
            "went": "go", "gone": "go",
            "came": "come",
            "took": "take",
            "saw": "see", "seen": "see",
            "knew": "know", "known": "know",
            "thought": "think",
            "said": "say",
            "told": "tell",
            "got": "get",
            "made": "make",
            "gave": "give",
            "found": "find",
            "became": "become",
            "left": "leave",
            "kept": "keep",
            "began": "begin",
            "shown": "show",
            "heard": "hear",
            "played": "play",
            "ran": "run",
            "moved": "move",
            "lived": "live",
            "believed": "believe",
            "brought": "bring",
            "wrote": "write",
            "sat": "sit",
            "stood": "stand",
            "lost": "lose",
            "paid": "pay",
            "met": "meet",
            "built": "build",
            "stayed": "stay",
            "fell": "fall",
            "sold": "sell",
            "sent": "send",
            "died": "die"
        ]

        if let lemma = irregularVerbs[lowerWord] {
            return lemma
        }

        if lowerWord.count > 3 && lowerWord.hasSuffix("ied") {
            return String(lowerWord.dropLast(3)) + "y"
        }

        if lowerWord.count > 3 && lowerWord.hasSuffix("ed") {
            return removeDoubledFinalConsonant(String(lowerWord.dropLast(2)))
        }

        if lowerWord.count > 4 && lowerWord.hasSuffix("ing") {
            return removeDoubledFinalConsonant(String(lowerWord.dropLast(3)))
        }

        if lowerWord.count > 3 && lowerWord.hasSuffix("ies") {
            return String(lowerWord.dropLast(3)) + "y"
        }

        if lowerWord.count > 3 && lowerWord.hasSuffix("es") {
            let base = String(lowerWord.dropLast(2))
            if ["sh", "ch", "x", "z", "ss"].contains(where: { base.hasSuffix($0) }) {
                return base
            }
            return String(lowerWord.dropLast(1))
        }

        if lowerWord.count > 2 && lowerWord.hasSuffix("s") {
            let base = String(lowerWord.dropLast(1))
            if !base.hasSuffix("s") {
                return base
            }
        }

        return lowerWord
    }

    private func removeDoubledFinalConsonant(_ word: String) -> String {
        guard word.count > 2, let last = word.last, let previous = word.dropLast().last else {
            return word
        }

        if last == previous && "bcdfghjklmnpqrstvwxyz".contains(last) {
            return String(word.dropLast())
        }

        return word
    }
}

private enum OcrPluginError: Error {
    case imageEncodingFailed
}

private extension CGImagePropertyOrientation {
    init(_ orientation: UIImage.Orientation) {
        switch orientation {
        case .up:
            self = .up
        case .upMirrored:
            self = .upMirrored
        case .down:
            self = .down
        case .downMirrored:
            self = .downMirrored
        case .left:
            self = .left
        case .leftMirrored:
            self = .leftMirrored
        case .right:
            self = .right
        case .rightMirrored:
            self = .rightMirrored
        @unknown default:
            self = .up
        }
    }
}

File: android/src/main/java/com/ocr/plugin/CapacitorPluginOcr.java
package com.ocr.plugin;

import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.net.Uri;
import android.util.Log;

import com.getcapacitor.JSArray;
import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import com.google.mlkit.vision.common.InputImage;
import com.google.mlkit.vision.text.Text;
import com.google.mlkit.vision.text.TextRecognition;
import com.google.mlkit.vision.text.TextRecognizer;
import com.google.mlkit.vision.text.latin.TextRecognizerOptions;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Locale;
import java.util.Map;
import java.util.Set;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

@CapacitorPlugin(name = "CapacitorPluginOcr")
public class CapacitorPluginOcr extends Plugin {

    private static final String TAG = "CapacitorPluginOcr";
    private static final Pattern ENGLISH_WORD_PATTERN = Pattern.compile("[a-zA-Z]+");

    private final Lemmatizer lemmatizer = new Lemmatizer();

    @PluginMethod
    public void checkPermissions(PluginCall call) {
        JSObject result = new JSObject();
        result.put("granted", true);
        call.resolve(result);
    }

    @PluginMethod
    public void requestPermissions(PluginCall call) {
        JSObject result = new JSObject();
        result.put("granted", true);
        call.resolve(result);
    }

    @PluginMethod
    public void recognizeEnglishText(PluginCall call) {
        String imagePath = call.getString("imagePath");
        if (imagePath == null || imagePath.trim().isEmpty()) {
            call.reject("Image path is required");
            return;
        }

        JSObject cropArea = call.getObject("cropArea");
        try {
            InputImage image = loadInputImage(imagePath, cropArea);
            TextRecognizer recognizer = TextRecognition.getClient(TextRecognizerOptions.DEFAULT_OPTIONS);

            recognizer.process(image)
                    .addOnSuccessListener(text -> call.resolve(processText(text)))
                    .addOnFailureListener(error -> {
                        Log.e(TAG, "OCR failed", error);
                        call.reject("OCR recognition failed: " + error.getMessage(), error);
                    })
                    .addOnCompleteListener(task -> recognizer.close());
        } catch (Exception error) {
            Log.e(TAG, "Failed to load image", error);
            call.reject("Could not load image: " + error.getMessage(), error);
        }
    }

    @PluginMethod
    public void cropImage(PluginCall call) {
        String imagePath = call.getString("imagePath");
        JSObject cropArea = call.getObject("cropArea");

        if (imagePath == null || imagePath.trim().isEmpty()) {
            call.reject("Image path is required");
            return;
        }

        if (cropArea == null) {
            call.reject("Crop area is required");
            return;
        }

        try {
            Bitmap original = decodeBitmap(resolveUri(imagePath));
            File croppedFile = cropBitmapToFile(original, cropArea);
            original.recycle();

            JSObject result = new JSObject();
            result.put("croppedImagePath", croppedFile.getAbsolutePath());
            call.resolve(result);
        } catch (Exception error) {
            Log.e(TAG, "Crop failed", error);
            call.reject("Crop failed: " + error.getMessage(), error);
        }
    }

    @PluginMethod
    public void startCropUI(PluginCall call) {
        call.reject("Interactive crop UI is not implemented. Use cropImage with a cropArea instead.");
    }

    private InputImage loadInputImage(String imagePath, JSObject cropArea) throws IOException {
        Uri uri = resolveUri(imagePath);

        if (cropArea == null) {
            return InputImage.fromFilePath(getContext(), uri);
        }

        Bitmap original = decodeBitmap(uri);
        File croppedFile = cropBitmapToFile(original, cropArea);
        original.recycle();
        return InputImage.fromFilePath(getContext(), Uri.fromFile(croppedFile));
    }

    private Uri resolveUri(String imagePath) {
        String path = imagePath.trim();

        if (path.startsWith("file://") || path.startsWith("content://")) {
            return Uri.parse(path);
        }

        return Uri.fromFile(new File(path));
    }

    private Bitmap decodeBitmap(Uri uri) throws IOException {
        InputStream stream = getContext().getContentResolver().openInputStream(uri);
        if (stream == null) {
            throw new IOException("Could not open image stream");
        }

        try (InputStream input = stream) {
            Bitmap bitmap = BitmapFactory.decodeStream(input);
            if (bitmap == null) {
                throw new IOException("Could not decode image");
            }
            return bitmap;
        }
    }

    private File cropBitmapToFile(Bitmap original, JSObject cropArea) throws IOException {
        int imgWidth = original.getWidth();
        int imgHeight = original.getHeight();

        double x = clamp(cropArea.optDouble("x", 0), 0, 1);
        double y = clamp(cropArea.optDouble("y", 0), 0, 1);
        double width = clamp(cropArea.optDouble("width", 1), 0, 1);
        double height = clamp(cropArea.optDouble("height", 1), 0, 1);

        int cropX = Math.min((int) Math.round(x * imgWidth), imgWidth - 1);
        int cropY = Math.min((int) Math.round(y * imgHeight), imgHeight - 1);
        int cropW = Math.max(1, Math.min((int) Math.round(width * imgWidth), imgWidth - cropX));
        int cropH = Math.max(1, Math.min((int) Math.round(height * imgHeight), imgHeight - cropY));

        Bitmap cropped = Bitmap.createBitmap(original, cropX, cropY, cropW, cropH);
        File outputFile = new File(getContext().getCacheDir(), "ocr_crop_" + System.currentTimeMillis() + ".jpg");

        try (FileOutputStream output = new FileOutputStream(outputFile)) {
            cropped.compress(Bitmap.CompressFormat.JPEG, 92, output);
        } finally {
            cropped.recycle();
        }

        return outputFile;
    }

    private JSObject processText(Text text) {
        Set<String> uniqueWords = new HashSet<>();
        JSArray rawWords = new JSArray();
        JSArray words = new JSArray();

        for (Text.TextBlock block : text.getTextBlocks()) {
            for (Text.Line line : block.getLines()) {
                for (Text.Element element : line.getElements()) {
                    Matcher matcher = ENGLISH_WORD_PATTERN.matcher(element.getText());

                    while (matcher.find()) {
                        String word = matcher.group().toLowerCase(Locale.ROOT);
                        if (word.length() <= 1) {
                            continue;
                        }

                        rawWords.put(word);
                        if (uniqueWords.add(word)) {
                            JSObject wordResult = new JSObject();
                            wordResult.put("word", word);
                            wordResult.put("confidence", 1.0);

                            String lemma = lemmatizer.lemmatize(word);
                            if (!lemma.equals(word)) {
                                wordResult.put("lemma", lemma);
                            }

                            words.put(wordResult);
                        }
                    }
                }
            }
        }

        JSObject result = new JSObject();
        result.put("words", words);
        result.put("rawWords", rawWords);
        return result;
    }

    private double clamp(double value, double min, double max) {
        return Math.max(min, Math.min(max, value));
    }

    private static class Lemmatizer {
        private static final Map<String, String> IRREGULAR_VERBS = new HashMap<>();

        static {
            IRREGULAR_VERBS.put("was", "be");
            IRREGULAR_VERBS.put("were", "be");
            IRREGULAR_VERBS.put("been", "be");
            IRREGULAR_VERBS.put("being", "be");
            IRREGULAR_VERBS.put("had", "have");
            IRREGULAR_VERBS.put("has", "have");
            IRREGULAR_VERBS.put("did", "do");
            IRREGULAR_VERBS.put("done", "do");
            IRREGULAR_VERBS.put("went", "go");
            IRREGULAR_VERBS.put("gone", "go");
            IRREGULAR_VERBS.put("came", "come");
            IRREGULAR_VERBS.put("took", "take");
            IRREGULAR_VERBS.put("saw", "see");
            IRREGULAR_VERBS.put("seen", "see");
            IRREGULAR_VERBS.put("knew", "know");
            IRREGULAR_VERBS.put("known", "know");
            IRREGULAR_VERBS.put("thought", "think");
            IRREGULAR_VERBS.put("said", "say");
            IRREGULAR_VERBS.put("told", "tell");
            IRREGULAR_VERBS.put("got", "get");
            IRREGULAR_VERBS.put("made", "make");
            IRREGULAR_VERBS.put("gave", "give");
            IRREGULAR_VERBS.put("found", "find");
            IRREGULAR_VERBS.put("became", "become");
            IRREGULAR_VERBS.put("left", "leave");
            IRREGULAR_VERBS.put("kept", "keep");
            IRREGULAR_VERBS.put("began", "begin");
            IRREGULAR_VERBS.put("shown", "show");
            IRREGULAR_VERBS.put("heard", "hear");
            IRREGULAR_VERBS.put("played", "play");
            IRREGULAR_VERBS.put("ran", "run");
            IRREGULAR_VERBS.put("moved", "move");
            IRREGULAR_VERBS.put("lived", "live");
            IRREGULAR_VERBS.put("believed", "believe");
            IRREGULAR_VERBS.put("brought", "bring");
            IRREGULAR_VERBS.put("wrote", "write");
            IRREGULAR_VERBS.put("sat", "sit");
            IRREGULAR_VERBS.put("stood", "stand");
            IRREGULAR_VERBS.put("lost", "lose");
            IRREGULAR_VERBS.put("paid", "pay");
            IRREGULAR_VERBS.put("met", "meet");
            IRREGULAR_VERBS.put("built", "build");
            IRREGULAR_VERBS.put("stayed", "stay");
            IRREGULAR_VERBS.put("fell", "fall");
            IRREGULAR_VERBS.put("sold", "sell");
            IRREGULAR_VERBS.put("sent", "send");
            IRREGULAR_VERBS.put("died", "die");
        }

        String lemmatize(String word) {
            String lowerWord = word.toLowerCase(Locale.ROOT);
            int length = lowerWord.length();

            if (IRREGULAR_VERBS.containsKey(lowerWord)) {
                return IRREGULAR_VERBS.get(lowerWord);
            }

            if (length > 3 && lowerWord.endsWith("ied")) {
                return lowerWord.substring(0, length - 3) + "y";
            }

            if (length > 3 && lowerWord.endsWith("ed")) {
                return removeDoubledFinalConsonant(lowerWord.substring(0, length - 2));
            }

            if (length > 4 && lowerWord.endsWith("ing")) {
                return removeDoubledFinalConsonant(lowerWord.substring(0, length - 3));
            }

            if (length > 3 && lowerWord.endsWith("ies")) {
                return lowerWord.substring(0, length - 3) + "y";
            }

            if (length > 3 && lowerWord.endsWith("es")) {
                String base = lowerWord.substring(0, length - 2);
                if (base.endsWith("sh") || base.endsWith("ch") || base.endsWith("x") || base.endsWith("z")
                        || base.endsWith("ss")) {
                    return base;
                }
                return lowerWord.substring(0, length - 1);
            }

            if (length > 2 && lowerWord.endsWith("s")) {
                String base = lowerWord.substring(0, length - 1);
                if (!base.endsWith("s")) {
                    return base;
                }
            }

            return lowerWord;
        }

        private String removeDoubledFinalConsonant(String word) {
            int length = word.length();
            if (length > 2) {
                char last = word.charAt(length - 1);
                char previous = word.charAt(length - 2);
                if (last == previous && isConsonant(last)) {
                    return word.substring(0, length - 1);
                }
            }
            return word;
        }

        private boolean isConsonant(char value) {
            return "bcdfghjklmnpqrstvwxyz".indexOf(value) >= 0;
        }
    }
}
